File: book8-cracking-the-coding-interview/preface-vi-big-o/Example9BinaryTreeSum.java
public class Example9BinaryTreeSum {

    class Node<T extends Number> {
        public Node<T> left;
        public Node<T> right;
        public T value;

        public Node() {
        }

        public Node(T value) {
            this.value = value;
        }

        public Node(Node<T> left, Node<T> right, T value) {
            this.left = left;
            this.right = right;
            this.value = value;
        }
    }

    /**
     * Sum all the nodes in the binary tree.
     * Time: O(n)
     * Memory: O(log(n))
     * 
     * @param <T>
     * @param node
     * @return
     */
    <T extends Number> int sum(Node<T> node) {
        if (node == null)
            return 0;
        return sum(node.left) + node.value.intValue() + sum(node.right);
    }

    void run() {
        var n1 = new Node<Integer>(1);
        var n2 = new Node<Integer>(2);
        var n3 = new Node<Integer>(3);
        var n4 = new Node<Integer>(4);
        var n5 = new Node<Integer>(5);
        n2.left = n1;
        n3.left = n2;
        n3.right = n4;
        n4.right = n5;

        System.out.println("Sum: " + sum(n3));
        System.out.println("SumMath:" + (5 * 6 / 2));
    }

    public static void main(String[] args) {
        new Example9BinaryTreeSum().run();
    }

}

File: book8-cracking-the-coding-interview/preface-vi-big-o/Example9BinaryTreeSum.py
class Node:
    def __init__(self, value=None, left=None, right=None):
        self.value = value
        self.left = left
        self.right = right

    def sum(self) -> int:
        """
        Calculates the sum of all nodes in the binary tree.
        Time: O(n)
        Memory: O(log(n))
        """
        return self._sum(self)

    def _sum(self, node=None):
        if node is None:
            return 0
        return self._sum(node.left) + node.value + self._sum(node.right)


def run():
    nodes = [Node(1), Node(2), Node(3), Node(4), Node(5)]
    nodes[2].left = nodes[1]
    nodes[2].right = nodes[3]
    nodes[1].left = nodes[0]
    nodes[3].right = nodes[4]
    print('Sum:', nodes[2].sum())


if __name__ == '__main__':
    run()
